request_logger.go

```diff
@@ -0,0 +1,121 @@
+	replacer := strings.NewReplacer(
+		"/", repl,
+		"\\", repl,
+		":", repl,
+		"*", repl,
+		"?", repl,
+		"\"", repl,
+		"<", repl,
+		">", repl,
+		"|", repl,
+	)
+	return replacer.Replace(model)
+}
+
+// logUpstreamRequest logs an HTTP request with the given ID and model name.
+// The prefix format is: [req] [id] [model]
+func logUpstreamRequest(logger *log.Logger, id, model string, req *http.Request) {
+	if logger == nil {
+		return
+	}
+
+	if reqDump, err := httputil.DumpRequest(req, true); err == nil {
+		logger.Printf("[req] [%s] [%s] %s", id, model, reqDump)
+	}
+}
+
+// logUpstreamResponse logs an HTTP response with the given ID and model name.
+// The prefix format is: [res] [id] [model]
+func logUpstreamResponse(logger *log.Logger, id, model string, resp *http.Response) {
+	if logger == nil {
+		return
+	}
+
+	if respDump, err := httputil.DumpResponse(resp, true); err == nil {
+		logger.Printf("[res] [%s] [%s] %s", id, model, respDump)
+	}
+}
+
+// logUpstreamError logs an error that occurred during request/response processing.
+// The prefix format is: [res] [id] [model] Error:
+func logUpstreamError(logger *log.Logger, id, model string, err error) {
+	if logger == nil {
+		return
+	}
+
+	logger.Printf("[res] [%s] [%s] Error: %v", id, model, err)
+}
+
+// createLoggingMiddleware creates a middleware function that logs requests and responses.
+// Logs are written to $baseDir/$provider/$model/$id.req.log and $baseDir/$provider/$model/$id.res.log
+// where baseDir is from cfg.UpstreamLoggingDir or os.TempDir() if not specified.
+// Returns nil if logging setup fails, logging errors via the provided logger.
+func createLoggingMiddleware(logger slog.Logger, cfg *ProviderConfig, provider, id, model string) func(*http.Request, func(*http.Request) (*http.Response, error)) (*http.Response, error) {
+	ctx := context.Background()
+	safeModel := SanitizeModelName(model)
+
+	baseDir := cfg.UpstreamLoggingDir()
+	if baseDir == "" {
+		baseDir = os.TempDir()
+	}
+
+	logDir := filepath.Join(baseDir, provider, safeModel)
+
+	// Create the directory structure if it doesn't exist
+	if err := os.MkdirAll(logDir, 0755); err != nil {
+		logger.Warn(ctx, "failed to create log directory", slog.Error(err), slog.F("dir", logDir))
+		return nil
+	}
+
+	reqLogPath := filepath.Join(logDir, fmt.Sprintf("%s.req.log", id))
+	resLogPath := filepath.Join(logDir, fmt.Sprintf("%s.res.log", id))
+
+	reqLogFile, err := os.OpenFile(reqLogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		logger.Warn(ctx, "failed to open request log file", slog.Error(err), slog.F("path", reqLogPath))
+		return nil
+	}
+
+	resLogFile, err := os.OpenFile(resLogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		reqLogFile.Close()
+		logger.Warn(ctx, "failed to open response log file", slog.Error(err), slog.F("path", resLogPath))
+		return nil
+	}
+
+	reqLogger := log.New(reqLogFile, "", log.LstdFlags)
+	resLogger := log.New(resLogFile, "", log.LstdFlags)
```

Comment:
The `log` package isn't really doing much for you here, I think you'd pretty much get the same effect by just `fmt.Fprintln`ing to the file directly.